rfd/0048-desktop-access-session-recording.md

```diff
@@ -0,0 +1,300 @@
+```
+
+On the backend, the log is downloaded from external storage, the events are
+parsed via an `AuditReader`, and finally written to the events channel. There is
+no attention paid to timing at this step. Events are written to the channel as
+soon as they are available and the process is naturally throttled by how fast
+the consumer reads from the channel.
+
+### Overview
+
+During a live session, desktop images are sent to the browser via a series of
+PNG frames encoded in a TDP message (message #2, "PNG frame"). At a high level,
+the session recording feature needs to archive these messages in the audit log
+so that they can be played back at a later time.
+
+During playback, the majority of the frontend code works exactly as it does
+during a live session - the PNGs are rendered on an HTML canvas.
+
+The major differences in playback mode are:
+
+- no user-input is captured or sent across the wire (mouse clicks, scroll wheel, etc)
+- playback features such as play/pause (at a minimum), and seek are desired
+
+#### Compatibility
+
+The session recording feature further increases the importance of
+backwards-compatible changes to the TDP spec, as newer versions of Teleport must
+be able to play back recordings captured by older versions.
+
+This means we must favor adding new messages over modifying existing messages,
+and retain support for deprecated messages in order to keep recordings valid.
+
+#### Recording
+
+Teleport session recordings are stored as an ordered sequence of protobuf-encoded
+audit events. Desktop session recording will work in the same way.
+
+In order to support a new type of session recording, a new audit event will be
+defined for capturing desktop session data. The `DesktopRecordingEvent` will
+implement the `AuditEvent` interface.
+
+```protobuf
+// DesktopRecordingEvent happens when a Teleport Desktop Protocol message
+// is captured during a Desktop Access Session.
+message DesktopRecordingEvent {
+    // Metadata is a common event metadata
+    Metadata Metadata = 1
+        [ (gogoproto.nullable) = false, (gogoproto.embed) = true, (gogoproto.jsontag) = "" ];
+
+    // Message is the encoded TDP message. It is not marshaled to JSON format
+    bytes Message = 2 [ (gogoproto.nullable) = true, (gogoproto.jsontag) = "-" ];
+
+    // DelayMilliseconds is the delay in milliseconds from the start of the session
+    int64 DelayMilliseconds = 3 [ (gogoproto.jsontag) = "ms" ];
+}
+```
+
+The `Message` field contains an artibrary TDP-encoded message, as defined in
+[RFD 0037](./0037-desktop-access-protocol.md). The vast majority of these events
+will contain TDP message 2 (PNG frame), but we will also capture:
+
+- message 1 - client screen spec: to know the initial size of the screen
+- message 6 - clipboard data: to indicate when the clipboard was used
+- message 3 - mouse move
+- message 4 - mouse button
+
+Of note: we will not be capturing other user input events (keyboard or mouse
```

Comment:
I should add a note here. We also decided not to capture keyboard input because it could contain sensitive info like passwords. We're not looking to be a keylogger.